azurerm/resource_arm_api_management.go

```diff
@@ -0,0 +1,672 @@
+				Computed: true,
+			},
+
+			"scm_url": {
+				Type:     schema.TypeString,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func resourceArmApiManagementServiceCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient).apiManagementServiceClient
+	ctx := meta.(*ArmClient).StopContext
+
+	log.Printf("[INFO] preparing arguments for AzureRM API Management creation.")
+
+	resGroup := d.Get("resource_group_name").(string)
+	name := d.Get("name").(string)
+	location := azureRMNormalizeLocation(d.Get("location").(string))
+	tags := d.Get("tags").(map[string]interface{})
+
+	var sku *apimanagement.ServiceSkuProperties
+	if skuConfig := d.Get("sku").([]interface{}); skuConfig != nil {
+		sku = expandAzureRmApiManagementSku(skuConfig)
+	}
+
+	properties := expandAzureRmApiManagementProperties(d, sku)
+
+	apiManagement := apimanagement.ServiceResource{
+		Location:          &location,
+		ServiceProperties: properties,
+		Tags:              expandTags(tags),
+		Sku:               sku,
+	}
+
+	createFuture, err := client.CreateOrUpdate(ctx, resGroup, name, apiManagement)
+	if err != nil {
+		return err
+	}
+
+	err = createFuture.WaitForCompletion(ctx, client.Client)
+	if err != nil {
+		return err
+	}
+
+	read, err := client.Get(ctx, resGroup, name)
+	if err != nil {
+		return err
+	}
+
+	if read.ID == nil {
+		return fmt.Errorf("Cannot read AzureRM Api Management %q (resource group %q) ID", name, resGroup)
+	}
+
+	d.SetId(*read.ID)
+
+	return resourceArmApiManagementServiceRead(d, meta)
+}
+
+func resourceArmApiManagementServiceRead(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient)
+	apiManagementClient := meta.(*ArmClient).apiManagementServiceClient
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+
+	resGroup := id.ResourceGroup
+	name := id.Path["service"]
+
+	ctx := client.StopContext
+	resp, err := apiManagementClient.Get(ctx, resGroup, name)
+
+	if err != nil {
+		if utils.ResponseWasNotFound(resp.Response) {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("Error making Read request on API Management Service %q (Resource Group %q): %+v", name, resGroup, err)
+	}
+
+	d.Set("name", name)
+	d.Set("resource_group_name", resGroup)
+
+	if location := resp.Location; location != nil {
+		d.Set("location", azureRMNormalizeLocation(*location))
+	}
+
+	if props := resp.ServiceProperties; props != nil {
+		d.Set("publisher_email", props.PublisherEmail)
+		d.Set("publisher_name", props.PublisherName)
+
+		d.Set("notification_sender_email", props.NotificationSenderEmail)
+		d.Set("created", props.CreatedAtUtc.Format(time.RFC3339))
+		d.Set("gateway_url", props.GatewayURL)
+		d.Set("gateway_regional_url", props.GatewayRegionalURL)
+		d.Set("portal_url", props.PortalURL)
+		d.Set("management_api_url", props.ManagementAPIURL)
+		d.Set("scm_url", props.ScmURL)
+		d.Set("static_ips", props.StaticIps)
+		d.Set("custom_properties", &props.CustomProperties)
```

Comment:
rather than setting these directly (which will result in a set error) - could we add a flatten function for this (as below) and convert this list to a `map[string]interface{}` rather than setting `map[string]string` here?